### DISCUSSION TITLE
Whether to ensure consistency for the 'ls'?

### DISCUSSION THREAD

**Discussion by Sallery-X** [juicedata/juicefs#1194] [Q&A] [answered]
**What would you like to be added**:

**Why is this needed**:

**Comment by davies** [comment, marked as answer]
To understand what happens behind `ls` command, you can `cat` the accesslog while execute `ls` command, here is what I got:

```
2021.12.21 21:34:06.707756 [uid:0,gid:0,pid:19694] opendir (1): OK [fh:77060] <0.000008>
2021.12.21 21:34:06.710774 [uid:0,gid:0,pid:19694] readdir (1,4096,0): OK (18) <0.002936>
2021.12.21 21:34:06.711609 [uid:0,gid:0,pid:19694] getattr (1): OK (1,[drwxrwxrwx:0040777,7,0,0,1610014882,1638287138,1638287138,4096]) <0.000704>
2021.12.21 21:34:06.711729 [uid:0,gid:0,pid:19694] readdir (1,4096,18): OK (0) <0.000003>
2021.12.21 21:34:06.711794 [uid:0,gid:0,pid:0] releasedir (1): OK <0.000002>
```  

It calls opendir, readdir, readdir, releasedir.  For the first readdir, JuiceFS will fetch all the children for the given directory in single call and save them for future readdir(). So the question will be is meta.Readdir() atomic or not?

the meta.Readdir() has two part: list all the entries and fetch all the attributes. they are different in three engines:
1) Readdir() is atomic in SQL engine.
2) the first part is atomic in TiKV, but the second is not, so you may see empty entry with ???.
3) The first part is atomic for small directory ( <10000 entries), NOT for larger one. Under heavy rename operation in same directory, you may miss an entry when it moved from later batch to previous [1]. 

[1] https://redis.io/commands/scan
